### ISSUE TITLE
Returned function with generic doesn't complain when provided type are incompatible

### ISSUE THREAD

**Issue by cavalier38** [microsoft/TypeScript#53610]
# Bug Report

With a generic function returning a function that has an argument base return type, which is later passed to another function.
It then is accepted by the later function even if the type used in the generic are incompatible.
For example a custom react hook which is passed down components.

When not the exact return type is used, but instead new type is written with the same structure.
Then the incomparibility in the type used as generic is detected.

### 🔎 Search Terms

returned
function
generic 

### 🕗 Version & Regression Information

Tested on 5.0.2 (Playground)
The syntax was accepted down to 4.7.4, all versions respond the same.
- This is the behavior in every version I tried (down to 4.7.4, below the syntax is not accepted, and I reviewed the FAQ for entries about generics

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?ts=5.0.2#code/GYVwdgxgLglg9mABCAzgUwCppVAPAcQwE8AHNAPgAoYwYoAuRQ0tASkQG8AoRRCBHIgDmaKADUAhgBsQaRAF5EBANJoiiNAA8oaMABMUiANZq4wJsTJUTRRvlVFWdy2gDa9tQF0F5Tj16IAE6iIIFINHSuNp4A3P4AvnG8wVChSCLi0rJx8VxcUCyIGPiikjJoBC6+igBKIWEuuAVkZsjoWDiVLOTkcVygkLAIiCQSKCiUGWWyjMWlWRUciBIAylCBNEKMOBtgQjGIYGhoeieMYCAAtgBGaIGI8eTsHLn94NDwSDo4lM-+-GBBJciC4FH4AssAHJXW6BRgAJgANP5eKt1ptGAByAAWaCkUjgmORvES-wEUGE83KYNQmGwUEowJcrCSIzGEymCxZiAA9DzEAB1bHqGCGKDY0V8OCXUawa5SNCIymZamSy6ilCbRCYo4nE6YgD8XHiQA)
[Playground link with working alternative](https://www.typescriptlang.org/play?#code/C4TwDgpgBAKg4hYA1AhgGwK4QDxxuCAPigF4pcBpCEKCAD2AgDsATAZygGtqB7AMyh4ChABTcQALkFUQASlLEhkANpwZAXQDcAKG18MTAMbAAljyZQMbCDAhtgufJFEmmJ4FKUR5Ab21QoQ3N7KABzRFRMCCl4CPQsR2FSKDFqeRJiPwCAgCdEDByLV3dlcS1-KABfHVz8wrC4qJ1K3X0jU3MoMBQ2NhFw5Hjo2ARBqOwfKBQAZWAc11Cpe3mmUM0oJggIFm2pJgwAWwAjCByqwl8WvQNjMwtGexFfCqCmEIOQJ2gyLOyUADlDicclIAEwAGgqARmcwWUgA5AALCBoNA8eGQgLVF7BYANMZYZJWGx2YAiD5fWQ1Lo9PoDSJYKlQAD0zKgAEk8SYOEEcnljGgaDtGMZtlBgDwoCcoK4ggduqYjmhoMi8gA6bSVIA)

### 💻 Code

```TypeScript
function useTest<GType>(init: GType) {
  const getValue = <GKey extends keyof GType>(key: GKey): GType[GKey] => {
    return init[key];
  };
  return getValue;
}

type TGetValue<GType> = ReturnType<typeof useTest<GType>>;

function pass(getValue: TGetValue<{ aString: string; needed: number }>) {}

function test() {
  const myType = {
    aNumber: 2,
    aString: 'hello',
  };
  const getValue = useTest(myType);
  pass(getValue);
}
```

### 🙁 Actual behavior

getValue is accepted as correct type for pass.

### 🙂 Expected behavior

getValue is missing 'needed' which is required by pass.

With alternative solution, changing TGetValue from ReturnType to a type on its own:
```TypeScript
type TGetValue<GType> = <GKey extends keyof GType>(key: GKey) => GType[GKey];
```
This report 'needed' is missing.
With this alternative code, when using version 3.3.3 it also accepts getValue like it does when using ReturnType.

**Comment by RyanCavanaugh** [comment]
Duplicate of #48070
